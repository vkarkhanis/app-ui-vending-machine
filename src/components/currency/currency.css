.bill {
    height: 40px;
    width: 50px;
    background-color: grey;
    position: relative;
  }

