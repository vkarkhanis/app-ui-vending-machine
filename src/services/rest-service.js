const BASE_URL = 'http://localhost:8080'

export const addAmount = async (value) => {

	try {
		return await updateAmount(`${BASE_URL}/vendingmachine/amount`, 'post', value);
	} catch (e) {
		console.error(e);
		throw e;
	}
	
}

export const increaseAmount = async (value, requestId) => {

	try {
		return await updateAmount(`${BASE_URL}/vendingmachine/amount?requestId=${requestId}`, 'put', value);
	} catch(err) {
		console.error(err);
		throw err;
	}
	
}

export const refundAmount = async (requestId) => {

	try {
		const resp = await fetch(`${BASE_URL}/vendingmachine/amount?requestId=${requestId}`);
		return await resp.json();
	} catch(err) {
		console.error(err);
		throw err;
	}
	
}

const updateAmount = async (url, method, value) => {
	
	const resp = await fetch(url, {
        method,
		headers: {
			"Content-type": "application/json"
		},
        body: JSON.stringify({amount: value}),
    });

    const vendingMachineResponse = await resp.json();
    return vendingMachineResponse;
}

export const fetchAllProducts = async () => {

	try {
		const resp = await fetch(`${BASE_URL}/vendingmachine/product`);
		const products = await resp.json();
		console.log('fetch all products', products)
		return products;
	} catch(err) {
		console.error(err);
		throw err;
	}
	
}

export const fetchProduct = async (productId, requestId) => {
	
	try {
		const resp = await fetch(`${BASE_URL}/vendingmachine/product/${productId}?requestId=${requestId}`);
		return await resp.json();
	} catch(err) {
		console.error(err);
		throw err;
	}
	
}