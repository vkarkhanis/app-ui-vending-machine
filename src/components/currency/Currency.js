import './currency.css';
import {addAmount, increaseAmount} from '../../services/rest-service';

export const Currency = (props) => {

    const onAmountAdded = async () => {

        const {onAmountAdded, requestId} = props;

		let vendingMachineResponse;
		
		if (requestId) {
			vendingMachineResponse = await increaseAmount(props.value, requestId);
		} else {
			vendingMachineResponse = await addAmount(props.value);	
		}
        
		if (vendingMachineResponse) {
			onAmountAdded && onAmountAdded(vendingMachineResponse);
		}
        
    }
	
    return (
        <button className={'bill'} onClick={onAmountAdded} disabled={props.disabled}>{props.value}</button>
    );
}