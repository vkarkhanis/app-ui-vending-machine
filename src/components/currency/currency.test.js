import React from 'react';
import ReactDOM from 'react-dom';
import { Currency } from './currency';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import * as restService from '../../services/rest-service';

jest.mock('../../services/rest-service');

beforeEach(() => {
    restService.addAmount = jest.fn().mockResolvedValue({amount: 20});
    restService.increaseAmount = jest.fn().mockResolvedValue({amount: 30});
})

afterEach(cleanup);

describe('Render Currency component test', () => {

    it('Currency component renders properly', () => {
        const divElem = document.createElement('div');
        ReactDOM.render(<Currency />, divElem);
        ReactDOM.unmountComponentAtNode(divElem);
    });
    
    it('Currency button name', () => {
        const { getByTestId } = render(<Currency disabled={false} value='Test' />);
        expect(getByTestId('vm-currency-btn-id')).toHaveTextContent('Test');
    });
    
    it('Match snapshot', () => {
        const tree = renderer.create(<Currency disabled={false} value='Test' onAmountAdded = {() => {}} />).toJSON();
        expect(tree).toMatchSnapshot();
    }); 

    it('Disable button', () => {
        const mockFn = jest.fn();
        const { getByTestId } = render(<Currency value='Test' disabled onAmountAdded={mockFn} />);
        expect(getByTestId('vm-currency-btn-id')).toBeDisabled();
    });

});

describe('Test Events', () => {

    it('Currency clicked', async () => {

        const mockFunction = jest.fn();

        const { getByTestId } = await waitFor(() => render(<Currency 
            disabled={false} value='Test' onAmountAdded={mockFunction} />));

        fireEvent.click(getByTestId('vm-currency-btn-id'));
        await waitFor(() => expect(mockFunction).toHaveBeenCalledTimes(1));
    });

});
