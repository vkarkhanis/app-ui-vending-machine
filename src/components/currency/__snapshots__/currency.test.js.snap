// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Currency component test Match snapshot 1`] = `
<button
  className="bill"
  data-testid="vm-currency-btn-id"
  disabled={false}
  onClick={[Function]}
>
  Test
</button>
`;
